Iceberg wasm blog post (#6260)
duckdb/duckdb-web: _posts/2025-12-12-iceberg-in-the-browser.md

For example, [Iceberg REST Catalogs](https://iceberg.apache.org/rest-catalog-spec/) provide these functionalities through a REST API.

There are two common ways to interact with Iceberg catalogs:

* The *client–server model,* where the compute part of the operation is delegated to a managed infrastructure (such as the cloud). Users can interact with the server by installing a local client or using a lightweight client such as a browser.
* The *client-is-the-server model,* where the user first installs the relevant libraries, and then performs queries directly on their machine.

Both models have their tradeoffs. In the *client–server model,* clients can be lightweight, and the server is the uniform point of access allowing for potential efficiencies thanks to scale. However, the infrastructure necessitates additional maintenance. In the *client-is-the-server model,* the query latency is lower, users can leverage local compute resources, and integrate with local inputs and external data sources happens at the user's level. However, requiring users to run computation locally means transferring part of the burden on your users (e.g., setting up dependencies).

> To fit these catalog models, the implementation of existing Iceberg engines fall into these categories: they can run locally as binaries or an organization can host them in their server infrastructure.

Reviewer: ```suggestion
A majority of Iceberg engine implementations fall into these interaction models: they can either be run locally by or by an organization that hosts them within their server infrastructure.
```

Reviewer 2: On this, I am not sure.
I think it's both more strong to say we are unique, and I also think that it's broadly true that the engine (that is different from the whole end to end beast) has to run somewhere, either locally or in a server.
Unsure what a better wording would be, I'll look it up this again later.

Reviewer: updated the comment. Going with "All Iceberg engine implementations" If someone gets mad, we apologize and update our blog post.

    TYPE iceberg,
    ENDPOINT_URL '⟨https://your-iceberg-endpoint⟩',
);

SELECT sum(value)
FROM db.table
WHERE other_column = '⟨some_value⟩';
```

> You can discover the full DuckDB-Iceberg extension feature set, including insert and update capabilities, in our [earlier blog post]({% post_url 2025-11-28-iceberg-writes-in-duckdb %}).

## Iceberg with DuckDB in the Browser

While setting up a local DuckDB installation is quite simple, opening a browser tab is even quicker.
Therefore, we asked ourselves: could we support the *client-is-the-server* model directly from within a browser tab?
This could provide zero-setup, no-infrastructure, properly serverless option for interacting with Iceberg catalogs.

![Iceberg with DuckDB-Wasm](/images/blog/iceberg-wasm/duckdb-iceberg-with-duckdb-wasm-dark.svg){: .darkmode-img }
![Iceberg with DuckDB-Wasm](/images/blog/iceberg-wasm/duckdb-iceberg-with-duckdb-wasm-light.svg){: .lightmode-img }

Luckily, DuckDB has a client that can run in any browser!
[DuckDB-Wasm]({% link docs/stable/clients/wasm/overview.md %}) is a WebAssembly port of DuckDB, which [supports loading of extensions]({% post_url 2023-12-18-duckdb-extensions-in-wasm %}).

> Interacting with a Iceberg REST Catalog requires a number of functionalities. First, the ability to talk to a REST API over HTTP(S). Second, the ability to read and write `avro` and `parquet` files on object storage. Third, negotiating authentication to access those resources on behalf of the user. All of these can be done also from a browser without the need to call a native component.

Reviewer: ```suggestion
Interacting with an Iceberg REST Catalog requires a number of functionalities; the ability to talk to a REST API over HTTP(S), the ability to read and write `avro` and `parquet` files on object storage, and finally, the ability to negotiate authentication to access those resources on behalf of the user. All of these can be done also from a browser without the need to call a native component.
```

To support these, we implemented the following high-level changes:

* In the core `duckdb` codebase, we redesigned HTTP interactions, so that extensions and clients have a uniform interface to the networking stack.
* In `duckdb-wasm`, we implemented such an interface, which in this case is a wrapper around the available JavaScript network stack.

> * In `duckdb-iceberg`, we routed all networking through the common HTTP interface, so that both in native and Wasm the same logic is executed.

Reviewer: ```suggestion
* In `duckdb-iceberg`, we routed all networking through the common HTTP interface, so that native DuckDB and DuckDB-wasm execute the same logic.
```

---
layout: post
title: "Iceberg in the Browser"
author: "Carlo Piovesan, Tom Ebergen, Gábor Szárnyas"
thumb: "/images/blog/thumbs/iceberg-in-the-browser.svg"
image: "/images/blog/thumbs/iceberg-in-the-browser.png"
excerpt: "DuckDB is the first end-to-end interface to Iceberg REST Catalogs within a browser tab. You can now read and write tables in Iceberg catalogs without needing to manage any infrastructure – directly from your browser!"
tags: ["deep dive"]
---

> Accessing an Iceberg REST Catalogs can be a complex operation requiring multiple moving parts.

Reviewer: ```suggestion
```

From the user's point of view, the choice of engine is transparent, and DuckDB is just one of many engines that the server could use in the background.
The *client-is-the-server* model is more interesting:
here, users [install a DuckDB client locally]({% link install/index.html %})
and use it through its SQL interface to query Iceberg catalogs.
For example:

```sql
CREATE SECRET test_secret (
    TYPE S3, 
    KEY_ID '⟨AKIAIOSFODNN7EXAMPLE⟩',
    SECRET '⟨wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY⟩'
);

ATTACH '⟨warehouse⟩' AS db (
    TYPE iceberg,
    ENDPOINT_URL '⟨https://your-iceberg-endpoint⟩',
);

SELECT sum(value)
FROM db.table
WHERE other_column = '⟨some_value⟩';
```

> You can discover the full DuckDB-Iceberg extension feature set, including insert and update capabilities, in our [earlier blog post]({% post_url 2025-11-28-iceberg-writes-in-duckdb %}).

## Iceberg with DuckDB in the Browser

While setting up a local DuckDB installation is quite simple, opening a browser tab is even quicker.
Therefore, we asked ourselves: could we support the *client-is-the-server* model directly from within a browser tab?

> This could provide zero-setup, no-infrastructure, properly serverless option for interacting with Iceberg catalogs.

Reviewer 2: ```suggestion
This could provide a zero-setup, no-infrastructure, properly serverless option for interacting with Iceberg catalogs.
```

To see a demo of serverless Iceberg analytics, visit our table visualizer at [`duckdb.org/visualizer`]({% link visualizer/index.html %}?iceberg).

TODO - final URL and video

## Access Your Own Data

As of today, this demo works with [Amazon S3 Tables]({% link docs/stable/core_extensions/iceberg/amazon_s3_tables.md %}). This has been implemented through a collaboration with the Amazon S3 Tables team.

You can now provide your own Iceberg warehouse and a set of credentials which allow reads from the catalog, metadata and data (policy [`AmazonS3TablesReadOnlyAccess`](https://us-east-1.console.aws.amazon.com/iam/home?region=us-east-2#/policies/details/arn%3Aaws%3Aiam%3A%3Aaws%3Apolicy%2FAmazonS3TablesReadOnlyAccess)).
Computations are fully local, and the credentials and warehouse ID are only sent to the catalog endpoint specified in your `Attach` command.
Inputs are translated to SQL, and added to the hash segment of the URL.

This means that:

* no sensitive data is handled or sent to `duckdb.org`
* computations are local, fully in your browser
* you can use the familiar SQL interface with the same code snippets can be run everywhere DuckDB runs

> * if you share the link, you might be sharing your credentials

Reviewer 2: Should we say that credentials we share a from a throw-away account and rotated?

* The *client-is-the-server model,* where the user first installs the relevant libraries, and then performs queries directly on their machine.

Obviously, both models have tradeoffs. In the client–server model, clients can be lightweight, and the server is the uniform point of access allowing for potential efficiencies thanks to scale. However, the infrastructure introduces additional maintenance requirements. In the client-is-the-server model, latency is lower, users can leverage local compute resources, and integrating with local inputs and external data sources happens at user level. However, requiring users to run computation locally means transferring part of the burden on your users (e.g., setting up dependencies).

The implementation of existing Iceberg engines follow these categories. Iceberg engines can run locally as binaries, or an organization can host them in their server infrastructure.

## Iceberg with DuckDB

![Iceberg with DuckDB](/images/blog/iceberg-wasm/iceberg-with-duckdb-dark.svg){: .darkmode-img }
![Iceberg with DuckDB](/images/blog/iceberg-wasm/iceberg-with-duckdb-light.svg){: .lightmode-img }

> DuckDB supports both Iceberg interaction models. For the client-is-the-server model, [install DuckDB locally](https://duckdb.org/install/) and use it as the client to query Iceberg using standard SQL. For example: 

Reviewer 2: ```suggestion
DuckDB supports both Iceberg interaction models, as engine run either in the *server* or as locally installable *client-is-the-server*.
From a user point of view, in the *client-server model* engine choice is transparent, while in the *client-is-the-server*, users will [install DuckDB locally](https://duckdb.org/install/) and use it through its SQL interface to query Iceberg Catalogs. For example: 
```

What does interacting with an Iceberg REST Catalog require? The ability to talk to a REST API over HTTP(S). The possibility of reading (or writing) `avro` and `parquet` files on object storage. Negotiating authentication to access those resources on behalf of the user.

At a high level, the changes required were:
* In DuckDB, we redesigned HTTP interactions, so that extensions and clients have an uniform interface to the networking stack.
* In DuckDB-Wasm, we implemented such an interface, that in this case is a wrapper around the available JavaScript network stack
* In DuckDB-Iceberg, we routed all networking through the common HTTP interface, so that both in native and Wasm the same logic is executed

The result is that you can now query Iceberg with DuckDB running in a browser tab:

![Iceberg with DuckDB-Wasm](/images/blog/duckdb-iceberg-with-duckdb-wasm-dark.svg){: .darkmode-img }
![Iceberg with DuckDB-Wasm](/images/blog/duckdb-iceberg-with-duckdb-wasm-light.svg){: .lightmode-img }

DuckDB-Iceberg in Wasm unlocks the analytical powers of DuckDB on your Iceberg catalog without infrastructure and without setup user side.

> Now you can access the same Iceberg Catalog using client–server, client-as-a-server, or properly serverless from the isolation of a single tab running on your users' browsers!

Reviewer 2: ```suggestion
![Iceberg with DuckDB-Wasm](/images/blog/duckdb-iceberg-with-duckdb-wasm-dark.svg){: .darkmode-img }
![Iceberg with DuckDB-Wasm](/images/blog/duckdb-iceberg-with-duckdb-wasm-light.svg){: .lightmode-img }

We looked at the available pieces, [DuckDB-Wasm](https://duckdb.org/docs/stable/clients/wasm/overview) is a WebAssembly port of DuckDB, which can run in any browser and [supports loading of extensions](https://duckdb.org/2023/12/18/duckdb-extensions-in-wasm).

What does interacting with an Iceberg REST Catalog require? The ability to talk to a REST API over HTTP(S). The possibility of reading (or writing) `avro` and `parquet` files on object storage. Negotiating authentication to access those resources on behalf of the user. All of the above can be done also from a Browser, no native component is actually required.

At a high level, the changes required were:
* In the core `duckdb` codebase, we redesigned HTTP interactions, so that extensions and clients have an uniform interface to the networking stack.
* In `duckdb-wasm`, we implemented such an interface, that in this case is a wrapper around the available JavaScript network stack
* In `duckdb-iceberg`, we routed all networking through the common HTTP interface, so that both in native and Wasm the same logic is executed

The result is that you can now query Iceberg with DuckDB running directly on a Browser.

Now you can access the same Iceberg Catalog using *client–server*, *client-as-a-server*, or properly serverless from the isolation of a single tab running on your browser!
```

Put first goal, then image/schema, then explaining how we are achieving it.

## Welcome to Serverless Iceberg Analytics

To see a demo of serverless Iceberg analytics, visit our table visualizer at [`duckdb.org/visualizer`](https://duckdb.org/visualizer/).

TODO - final URL and video

## Access Your Own Data

As of today, this demo works with [Amazon S3 Tables]({% link docs/stable/core_extensions/iceberg/amazon_s3_tables.md %}). This has been implemented through a collaboration with the Amazon S3 Tables team.

You can now provide your own Iceberg warehouse and a set of credentials that allows reads from the catalog, metadata and data (policy [`AmazonS3TablesReadOnlyAccess`](https://us-east-1.console.aws.amazon.com/iam/home?region=us-east-2#/policies/details/arn%3Aaws%3Aiam%3A%3Aaws%3Apolicy%2FAmazonS3TablesReadOnlyAccess)).

> Computations are fully local, and the credentials and warehouse ID are only sent towards the Catalog endpoint you specified.

Reviewer: ```suggestion
Computations are fully local, and the credentials and warehouse ID are only sent to the catalog endpoint specified in your `Attach` command.
```

tags: ["extension"]
---

In this blog post, we describe the state of Iceberg analytics and introduce a new possibility: interacting with an Iceberg REST Catalog can be as simple as navigating to an URL, no further setup or installs required.

## Interaction Models for Iceberg Catalogs

![Iceberg analytics today](/images/blog/iceberg-wasm/iceberg-analytics-today-dark.svg){: .darkmode-img }
![Iceberg analytics today](/images/blog/iceberg-wasm/iceberg-analytics-today-light.svg){: .lightmode-img }

Iceberg is an Open Table Format, that is to say a database stored as a collection of Parquet files (on object storage), metadata on those files (also static files on object storage) and a REST API to orchestrate data and metadata.

There are two common ways to interact with Iceberg catalogs:
* The *client–server model,* where the compute part of the operation is delegated to a managed infrastructure – such as the cloud. Users can interact with the server by installing a native client or a lightweight client such as a browser.
* The *client-is-the-server model,* where the user first installs the relevant libraries, and then performs queries directly on their machine.

> Obviously, both models have tradeoffs. In the client–server model, clients can be lightweight, and the server is the uniform point of access allowing for potential efficiencies thanks to scale. However, the infrastructure introduces additional maintenance requirements. In the client-is-the-server model, latency is lower, users can leverage local compute resources, and integrating with local inputs and external data sources happens at user level. However, requiring users to run computation locally means transferring part of the burden on your users (e.g., setting up dependencies).

Reviewer 2: ```suggestion
Both models have tradeoffs. In the client–server model, clients can be lightweight, and the server is the uniform point of access allowing for potential efficiencies thanks to scale. However, the infrastructure introduces additional maintenance requirements. In the client-is-the-server model, latency is lower, users can leverage local compute resources, and integrating with local inputs and external data sources happens at user level. However, requiring users to run computation locally means transferring part of the burden on your users (e.g., setting up dependencies).
```
